scripts/lib/compareTree.js

```diff
@@ -0,0 +1,150 @@
+const parents = require("unist-util-parents");
+const { selectAll } = require("unist-util-select");
+
+const getParents = (node) => {
+  var chain = [];
+  while (node) {
+    node.data.title && chain.unshift(node.data.title);
+    node = node.parent;
+  }
+  return chain;
+};
+
+// find the first parent text id to make legifrance links later
+const getParentTextId = (node) => {
+  let id;
+  node = node.parent;
+  while (node) {
+    if (
+      node.data &&
+      node.data.id &&
+      node.data.id.match(/^(KALI|LEGI)TEXT\d+$/)
+    ) {
+      id = node.data.id;
+      break;
+    }
+    node = node.parent;
+  }
+  return id || null;
+};
+
+// find the root text id to make legifrance links later
+const getRootId = (node) => {
+  let id;
+  while (node) {
+    id = node.data.id;
+    node = node.parent;
+  }
+  return id || null;
+};
+
+const addContext = (node) => ({
+  ...node,
+  parents: getParents(node),
+  textId: getParentTextId(node) || null,
+  rootId: getRootId(node) || null,
+});
+
+// dont include children in final results
+const stripChildren = (node) => node; //({ children, ...props }) => props;
+
+// return diffed articles nodes
+const compareArticles = (tree1, tree2, comparator) => {
+  const parentsTree1 = parents(tree1);
+  const parentsTree2 = parents(tree2);
+
+  // all articles from tree1
+  const articles1 = selectAll("article", parentsTree1).map(addContext);
+  const articles1cids = articles1
+    .map((a) => a && a.data && a.data.cid)
+    .filter(Boolean);
+  // all articles from tree2
+  const articles2 = selectAll("article", parentsTree2).map(addContext);
+  const articles2cids = articles2
+    .map((a) => a && a.data && a.data.cid)
+    .filter(Boolean);
+
+  // new : articles in tree2 not in tree1
+  const newArticles = articles2.filter(
+    (art) => art && art.data && !articles1cids.includes(art.data.cid)
+  );
+  const newArticlesCids = newArticles.map((a) => a.data.cid);
+
+  // supressed: articles in tree1 not in tree2
+  const missingArticles = articles1.filter(
+    (art) => art && art.data && !articles2cids.includes(art.data.cid)
+  );
+
+  // modified : articles with modified texte
+  const modifiedArticles = articles2.filter(
+    (art) =>
+      art &&
+      art.data &&
+      // exclude new articles
+      !newArticlesCids.includes(art.data.cid) &&
+      articles1.find(
+        // same article, different texte
+        (art2) =>
+          art2 &&
+          art2.data &&
+          art2.data.cid === art.data.cid &&
+          comparator(art, art2)
+      )
+  );
+
+  // all sections from tree1
+  const sections1 = selectAll("section", parentsTree1.children).map(addContext);
+
+  const sections1cids = sections1.map((a) => a.data.cid);
+
+  // all sections from tree2
+  const sections2 = selectAll("section", parentsTree2.children).map(addContext);
```

Comment:
@revolunet made some change here to only look to children root node to avoid having trouble it some root data change :D